### PR TITLE
Refactor ApplyContextCore

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
## What have you changed? (mandatory)

Refactor ApplyContext. This is a preparation for [apply pool](https://github.com/pingcap/tikv/pull/2960). 

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Refactor

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

Apply pool https://github.com/pingcap/tikv/pull/2960

### CODE REVIEW

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -341,21 +354,23 @@ impl<'a> ApplyContextCore<'a> {
                     panic!("failed to write to engine: {:?}", e);
                 });
         }
-        for cbs in self.cbs.drain(..) {
-            cbs.invoke_all(self.host);
+        for mut cbs in self.cbs.drain(..) {
```
Comment (hicqu) [inline src/raftstore/store/worker/apply.rs]: `mut` is not necessary?

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -564,7 +551,6 @@ impl ApplyDelegate {
         // others will be saved as a normal entry with no data, so we must re-propose these
         // commands again.
         apply_ctx.committed_count += committed_entries.len();
```
Comment (hicqu) [inline src/raftstore/store/worker/apply.rs]: Seems `committed_count` is updated twice?

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -2062,12 +2048,11 @@ impl Runner {
...
-        let mut core = ApplyContextCore::new(self.host.as_ref(), self.importer.as_ref())
-            .apply_res_capacity(applys.len())
-            .use_delete_range(self.use_delete_range)
-            .enable_sync_log(self.sync_log);
+        let mut merged_regions: Vec<u64> = vec![];
+        let mut apply_res: Vec<ApplyRes> = vec![];
+        let mut committed_count = 0;
         for apply in applys {
-            if apply.entries.is_empty() || core.merged_regions.contains(&apply.region_id) {
+            if apply.entries.is_empty() || merged_regions.contains(&apply.region_id) {
```
Comment (hicqu) [inline src/raftstore/store/worker/apply.rs]: Seems `merged_regions` is always empty.

**Comment by hicqu** [comment]
rest LGTM.

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -2081,8 +2065,24 @@ impl Runner {
             delegate.term = apply.term;
 
             {
-                let mut ctx = ApplyContext::new(&mut core, &mut self.delegates);
+                let mut ctx = ApplyContext::new(
+                    Arc::clone(&self.host),
+                    Arc::clone(&self.importer),
+                    &mut self.delegates,
+                ).use_delete_range(self.use_delete_range)
+                    .enable_sync_log(self.sync_log);
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: Forgot to set `ApplyContext::apply_res_capacity`?

File: src/raftstore/store/worker/apply.rs (hunk 2 above)
Comment (Hoverbear) [reply src/raftstore/store/worker/apply.rs]: @zhangjinpeng1987 Was this resolved?

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -274,20 +282,20 @@ impl<'a> ApplyContextCore<'a> {
...
-    pub fn enable_sync_log(mut self, eanbled: bool) -> ApplyContextCore<'a> {
-        self.enable_sync_log = eanbled;
-        self
+    pub fn push_exec_result(&mut self, res: ExecResult) {
+        self.exec_results.push(res);
     }
 
-    pub fn apply_res_capacity(mut self, cap: usize) -> ApplyContextCore<'a> {
-        self.apply_res = Vec::with_capacity(cap);
+    pub fn enable_sync_log(mut self, eanbled: bool) -> Self {
```
Comment (Hoverbear) [inline src/raftstore/store/worker/apply.rs]: You have a spelling mistake here. `enabled`

File: src/raftstore/store/worker/apply.rs [hunk 6]
```diff
@@ -341,21 +349,23 @@ impl<'a> ApplyContextCore<'a> {
...
 
     /// Finish applys for the delegate.
-    pub fn finish_for(&mut self, delegate: &mut ApplyDelegate, results: Vec<ExecResult>) {
+    pub fn finish_for(&mut self, delegate: &mut ApplyDelegate) {
         if !delegate.pending_remove {
             delegate.write_apply_state(self.wb_mut());
         }
         self.commit_opt(delegate, false);
+        let mut exec_results = vec![];
+        mem::swap(&mut exec_results, &mut self.exec_results);
```
Comment (Hoverbear) [inline src/raftstore/store/worker/apply.rs]: :+1:

File: src/raftstore/store/worker/apply.rs (hunk 2 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/apply.rs]: Done

**Pull request merged**